### ISSUE TITLE
JDWP-based Debugging Support

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#5648]
### TL;DR

Provide support for [JDWP](https://docs.oracle.com/javase/7/docs/technotes/guides/jpda/jdwp-spec.html)-based debugging so that users can debug native executables generated with Native Image using their preferred Java debugger.

### Goals

The [DebugInfo](https://www.graalvm.org/latest/reference-manual/native-image/debugging-and-diagnostics/DebugInfo/) feature of Native Image allows users to debug their native executables using GDB on the Java source level and works with many IDEs that provide GDB integrations. While this is crucial for debugging low-level problems in user applications or Native Image internals, many Java developers are unfamiliar with debugging of optimized native code.

To allow Java developers to use their preferred Java debugger, we plan to add support for higher-level debugging based on the JDWP. This way, optimizations, that are observable on the GDB level, can be hidden from users. Moreover, this would also make it possible to support *evaluate expression* and other higher-level debugging features in the future.

### Non-goals

* Stop work on GDB-based debugging infrastructure such as #5047
* Build and provide any IDE-specific plugins for Native Image debugging

**Comment by linghengqian** [comment]
- How is this issue related to https://github.com/oracle/graal/pull/10344? I'm curious why https://github.com/oracle/graal/pull/10344 is not linked to this issue.
- Or am I missing something, like there exists a Jetbrains YouTrack issue tracking IntelliJ IDEA support for `buildArg` of `-H:+JDWP`?

**Comment by fniephaus** [comment]
Hi @linghengqian,

This issue is very much related to https://github.com/oracle/graal/pull/10344, we just haven't added it to the issue description yet. We only opened the PR earlier this week and it got merged over night.

This upcoming feature will work with IntelliJ IDEA but it will also work in other Java IDEs that make use of JDWP. Note, however, that the feature is still in development and there are a couple of things that need to be done before we'll start advertising this to end users. If you like to try this out, you can use a [recent EA build of Oracle GraalVM](https://github.com/graalvm/oracle-graalvm-ea-builds/releases) and add `-H:+JDWP` to your list of Native Image build args.

**Comment by korkutkose** [comment]
Awesome news, thanks for the hard work. A question; Will this be backported to jdk17 or jdk21? Or this is considered as a new feature and will only be available jdk24 and onwards?

**Comment by fniephaus** [comment]
Thanks, @korkutkose! And no, this is a new feature and will not be backported to GraalVM for JDK 17 or 21.

**Comment by alina-yur** [comment]
Status update: we added experimental support in GraalVM 25 via the `-H:+JDWP` option, see [the documentation](https://www.graalvm.org/jdk25/reference-manual/native-image/debugging-and-diagnostics/JDWP/) to get started. Further work is ongoing, expect more updates in upcoming releases.
